### PR TITLE
UTF-8 encoded fields

### PR DESCRIPTION
**Pull request by Pingvin235**
Dear contributors, please consider if the attached changes are useful in the base repository. Test sets of course may be extended. Thanks for the great product!

### CODE REVIEW

**Comment by chibenwa** [comment]
Thanks for the contrib!

I have a naive question. Specification wise, shouldn't UTF-8 chars be quoted? Shouldn't mime messages be using US-ASCII charset?

Do you have at hand:
 - Specification for mime message in UTF-8 ?
 - Behaviour of other product / libs passing UTF-8 directly without encoding it?

**Comment by Pingvin235** [comment]
Regarding directly UTF-8 encoding.
* I have E-Mail samples from customers which are correctly processed in other clients. And they want to see them in BGERP as well.
* Found that partially (only for fields) such encoding supported in currently used JavaMail with activation `mail.mime.allowutf8` parameter.
https://javaee.github.io/javamail/docs/api/index.html?javax/mail/internet/package-summary.html
Unfortunately have no glue is that specified somewhere or not.

Q/Base64 encoding in fields as far as I know is completely standard, supported in JavaMail and often seen in sources.  
https://github.com/apache/james-mime4j/pull/55/files#diff-038c4f63b17523939c4c5cae5e32c1dc8d35a5cecb9c2093941382984e4891afR121
https://github.com/apache/james-mime4j/pull/55/files#diff-110316b99e33390efddd1d52a7b30e9fc098b2e93fa45e84a1ad07743151a9c7R442

**Comment by chibenwa** [comment]
`Q/Base64 encoding in fields as far as I know is completely standard, supported in JavaMail and often seen in sources.`

There is methods to decode encode Q in Mime4j though I admit the API to be unfriendly and hard to discover.

**Comment by chibenwa** [comment]
Can you open an issue there https://issues.apache.org/jira/projects/MIME4J/issues/MIME4J-250?filter=allopenissues ?

**Comment by Pingvin235** [comment]
An issue per case I guess?

File: dom/src/test/java/org/apache/james/mime4j/message/HeaderImplTest.java [hunk 1]
```diff
@@ -70,7 +75,7 @@ public void testWriteSpecialCharacters() throws Exception {
         byte[] b = outstream.toByteArray();
         ByteArrayBuffer buf = new ByteArrayBuffer(b.length);
         buf.append(b, 0, b.length);
-        String s = ContentUtil.decode(buf);
+        String s = ContentUtil.decode(StandardCharsets.UTF_8, buf);
```
Comment (chibenwa) [inline dom/src/test/java/org/apache/james/mime4j/message/HeaderImplTest.java:78]: I would have the charset as a second argument.

**Comment by chibenwa** [comment]
Ok, you basically want Q encoding decoded everywhere by default.

**Comment by chibenwa** [comment]
I would separate the UTF-8 topic from the Q encoding topic.

File: dom/src/test/java/org/apache/james/mime4j/message/HeaderImplTest.java (hunk 1 above)
Comment (Pingvin235) [reply dom/src/test/java/org/apache/james/mime4j/message/HeaderImplTest.java:78]: The method was already there, should it be changed?

File: core/src/main/java/org/apache/james/mime4j/stream/RawFieldParser.java [hunk 2]
```diff
@@ -191,6 +195,12 @@ public String parseToken(final ByteSequence buf, final ParserCursor cursor, fina
      *  is not delimited by any character.
      */
     public String parseValue(final ByteSequence buf, final ParserCursor cursor, final BitSet delimiters) {
+        if (!CharsetUtil.isASCII(buf)) {
+            String value = parseUtf8Filename(buf);
```
Comment (chibenwa) [inline core/src/main/java/org/apache/james/mime4j/stream/RawFieldParser.java:199]: So every time we parse a raw field we look ahead to see if it is a finename?

This do not look performant to me. I can see in the the future more and more `if` being added in there.

The mime4J logic tends more to be to delay parsing.

I would way prefer having helper methods in the ContentDisposition field (and other fields as well) doing Q decoding just if the user wants it (`contentDisposition.getDecodedFilename()`)

File: dom/src/test/java/org/apache/james/mime4j/message/HeaderImplTest.java (hunk 1 above)
Comment (chibenwa) [reply dom/src/test/java/org/apache/james/mime4j/message/HeaderImplTest.java:78]: Then no I think

**Comment by chibenwa** [comment]
Did you evaluate the performance impact of your changes (`/benchmark` for instance?)

**Comment by Pingvin235** [comment]
> Did you evaluate the performance impact of your changes (`/benchmark` for instance?)
Not yet. Will do that after separating them.

File: core/src/main/java/org/apache/james/mime4j/stream/RawFieldParser.java (hunk 2 above)
Comment (Pingvin235) [reply core/src/main/java/org/apache/james/mime4j/stream/RawFieldParser.java:199]: I guess you meant UTF-8 decoding here?

**Comment by chibenwa** [comment]
Any update on the benchmark front?

**Comment by Pingvin235** [comment]
> Any update on the benchmark front?

Not yet. Sorry, missed your question. For now I'm solving all the findings for our project in a separated fork's branch. But going to return everything finally to the upstream.

**Comment by Pingvin235** [comment]
Running `mvn test` in `/benchmark` directory.

Branch **master**.
```
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  3.534 s
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  3.646 s
```
Branch **utf8-encoded-fields**
```
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  3.350 s
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  3.535 s
```

So, there is no noticeable performance impact so far.

**Review by chibenwa — APPROVED** [review]

**Comment by chibenwa** [comment]
Do you think we can claim support for https://datatracker.ietf.org/doc/html/rfc6532 `Internationalized Email Headers` ?

I read the RFC, it seems to be it!

**Comment by Pingvin235** [comment]
Looks quite similar. But the support was already in code, I am just fixing some special cases, met by our customers.

**Review by Arsnael — APPROVED** [review]
Nice contribution in general, thank you :)

**Comment by Pingvin235** [comment]
My pleasure! What should be more done with the PR from my side in order to finish it and accept the changes?

**Comment by Arsnael** [comment]
Honestly I almost merged it myself.

But I ain't sure if @chibenwa intends of having a mention somewhere that we officially claim support the corresponding RFC?

**Comment by chibenwa** [comment]
Somewhere being https://github.com/apache/james-mime4j/blob/master/src/site/xdoc/start/index.xml

We also have that on James side: https://github.com/apache/james-project/blob/master/server/apps/distributed-app/docs/modules/ROOT/pages/architecture/implemented-standards.adoc#message-formats

**Pull request merged**